src/content/drafts/sourcecode-to-tokens-AST.md

```diff
@@ -0,0 +1,103 @@
+---
+layout: post
+title: How a parser translates text into tokens and then into an abstract syntax tree
+teaser: "You will learn how the source code in any proramming language is converted to token and then parsed to generate an AST. The AST can then be traversed to write and build custom rules in ESLint."
+image: 
+authors:
+  - Gayathri Krishnaswamy
+categories:
+  - Blog post
+tags:
+  - Lexical analysis
+  - Syntax analysis
+  - Abstract Syntax Tree
+  - AST
+  - Tokens
+  - ESLint rules
+---
+##  Introduction
+[ESLint](https://eslint.org/docs/latest/) is an open-source pluggable JavaScript linter that uses the Abstract Syntax Tree (AST) of your source code to identify and report patterns found in ECMAScript/JavaScript code. The tool also provides an option to write and configure your own rules to enforce a particular coding style that is relevant only to your project.
+
+In this article, first, we'll see on a high-level how your source code is converted into multiple tokens and is then parsed to produce an AST that represents the syntactic structure of your source code. Next, we'll understand how ESLint leverages the capability of AST and helps you build your own custom rules.
+
+Before we jump into details, let's understand a couple of key terminologies:
+
+- Compiler - A [compiler](https://en.wikipedia.org/wiki/Compiler) is a special program that converts the programming language’s source code into machine-readable code that targets specific operating systems and computer architectures. 
```

Comment:
If we are going to introduce compilers as part of this blog, then I think we need to make it clear a compiler uses a parser.

Reply:
I have modified the Introduction and removed references to compilers. Please let me know if this is ok.

A parser is a program that takes input in the form of sequential source program instructions and breaks it into smaller pieces (tokens) to generate statements and expressions. The parser then represents the statements and expressions in the form of a parse tree or an Abstract Syntax Tree (AST) that depicts the syntactic structure of your source code.

In this tutorial, first, you'll learn on a high-level how your source code is broken down into multiple tokens and is then converted to build an AST. Next, you'll learn how [ESLint](https://eslint.org/docs/latest/) leverages the capability of AST and helps you build your own custom rules.

The main steps involved in how the parsers process the source code are:

Step 1: Lexical analysis

Step 2: Syntax analysis

Reply:
The audience is someone who uses ESLint and wants to understand more about how it works. Generally, I think what you have here is a good introduction. One thing that is worth calling out is that parsers do not all necessary create an AST. An AST is created only when someone wants to be able to inspect the overall structure of a program. Parsers written for other purposes, such as immediate conversion into another format, may not create an AST at all.

Reply:
Okay. I have modified the Introduction as follows:

A parser is a program that takes input in the form of sequential source program instructions and breaks it into smaller pieces (tokens) to generate statements and expressions. The parser then represents the statements and expressions in the form of a parse tree or an Abstract Syntax Tree (AST) that depicts the syntactic structure of your source code.

:information_source: Not all parsers create an AST. An AST is required only to inspect the overall structure of the program. Parsers written for other purposes, such as immediate conversion into another format, may not create an AST at all.

In this tutorial, first, you'll learn on a high-level how your source code is broken down into multiple tokens and is then converted to an AST. Next, you'll learn how [ESLint](https://eslint.org/docs/latest/) leverages the capability of AST and helps you build your own custom rules. This tutorial is intended for those who use ESLint and wants to understand more about how the tool works. It assumes that you have basic familiarity with:

- Computer programming
- Javascript coding
- ESLint tool

By the end of this tutorial, you’ll be able to:


-  Understand how the source code is converted to AST.
- Generate AST for your source program using AST explorer.
- Understand how to traverse the AST nodes to write custom rules in ESLint.